### PR TITLE
Fix diff around default and known settings

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

We need to mimic the behavior from viper around default and known settings.

- Fix the case where the user would `BindEnv` after setting a default and would override the default with `nil`:
```
config.SetDefault("a", 1234)
config.BindEnv("a")
```
- Fix intermediate key not being known: `SetKnown("a.b.c")` should set `a`, `a.b` and `a.b.c` as known.
- Fix AllSettings returning settings without values from `BindEnv`. This is done by adding a proper `schema` to differentiate settings with and without values. This answer the core difference between`BindEnv` add  `SetDefault`. Once we enforce a default for all settings we will be able to remove this difference. but for now we must mimic `Viper` default.
- Fix the config not using search paths to look for `datadog.yaml`
- Fix `IsSet` returning false when given an inner node setting path
- Fix `proxy` setting initialization: this settings was initialize as a leaf but use as an inner node by viper.
- Reproduce behavior from Viper where the `Get` method would infer the return var type from the defaults.

### Describe how you validated your changes

Functionality is currently guarded behind a env var, so there should be no functional change. To test the behavior, use the DD_CONF_NODETREEMODEL env var, but be warned that there are likely bugs.

### CODE REVIEW

File: pkg/config/teeconfig/teeconfig.go [hunk 1]
```diff
@@ -156,11 +164,9 @@ func (t *teeConfig) AllKeysLowercased() []string {
...
 
-func compareResult(key, method string, base, compare interface{}) interface{} {
+func (t *teeConfig) compareResult(key, method string, base, compare interface{}) interface{} {
 	if !reflect.DeepEqual(base, compare) {
-		_, file, line, _ := runtime.Caller(2)
-		fileParts := strings.Split(file, "DataDog/datadog-agent/")
-		log.Warnf("difference in config: %s(%s) -> base: %v | compare %v from %s:%d", method, key, base, compare, fileParts[len(fileParts)-1], line)
+		log.Warnf("difference in config: %s(%s) -> base[%s]: %#v | compare[%s] %#v | from %s", method, key, t.baseline.GetSource(key), base, t.compare.GetSource(key), compare, getLocation(2))
 	}
 	return compare
```
Comment (dustmop) [inline pkg/config/teeconfig/teeconfig.go:171]: This seems like a bug: `compareResult` is returning `compare` which is the nodetreemodel value. It should be returning `base` instead, which is the value from viper.

File: pkg/config/teeconfig/teeconfig.go [hunk 2]
```diff
@@ -169,22 +175,22 @@ func compareResult(key, method string, base, compare interface{}) interface{} {
 func (t *teeConfig) Get(key string) interface{} {
 	base := t.baseline.Get(key)
 	compare := t.compare.Get(key)
-	return compareResult(key, "Get", base, compare)
+	return t.compareResult(key, "Get", base, compare)
```
Comment (dustmop) [inline pkg/config/teeconfig/teeconfig.go:178]: Which means here the future value is being returned when teeconfig is enabled.

File: pkg/config/setup/config.go [hunk 3]
```diff
@@ -1037,7 +1037,13 @@ func agent(config pkgconfigmodel.Setup) {
 	config.BindEnv("dd_url", "DD_DD_URL", "DD_URL")
 	config.BindEnvAndSetDefault("app_key", "")
 	config.BindEnvAndSetDefault("cloud_provider_metadata", []string{"aws", "gcp", "azure", "alibaba", "oracle", "ibm"})
-	config.SetDefault("proxy", nil)
+	config.SetDefault("proxy.http", nil)
+	config.SetDefault("proxy.https", nil)
+	config.SetDefault("proxy.no_proxy", nil)
+	//config.SetDefault("proxy.http", "")
```
Comment (dustmop) [inline pkg/config/setup/config.go:1043]: Remove these commented out lines.

File: pkg/config/nodetreemodel/read_config_file.go [hunk 4]
```diff
@@ -123,8 +134,8 @@ func loadYamlInto(defaults InnerNode, dest InnerNode, data map[string]interface{
 		key = strings.ToLower(key)
 		curPath := path + key
 
-		// check if the key is know in the defaults
-		defaultNode, err := defaults.GetChild(key)
+		// check if the key is know in the schema
+		defaultNode, err := schema.GetChild(key)
```
Comment (dustmop) [inline pkg/config/nodetreemodel/read_config_file.go:138]: How about also renaming `defaultNode`.

File: pkg/config/nodetreemodel/read_config_file.go [hunk 5]
```diff
@@ -76,10 +83,14 @@ func (c *ntmConfig) readInConfig(filePath string) error {
 
 func (c *ntmConfig) readConfigurationContent(target InnerNode, content []byte) error {
 	var obj map[string]interface{}
-	if err := yaml.Unmarshal(content, &obj); err != nil {
-		return err
+
+	if strictErr := yaml.UnmarshalStrict(content, &obj); strictErr != nil {
+		log.Errorf("warning reading config file: %v\n", strictErr)
+		if err := yaml.UnmarshalStrict(content, &obj); err != nil {
```
Comment (dustmop) [inline pkg/config/nodetreemodel/read_config_file.go:89]: Why call this a second time? Is it supposed to be calling `yaml.Unmarshal` instead?

File: pkg/config/nodetreemodel/read_config_file.go [hunk 6]
```diff
@@ -19,7 +20,12 @@ import (
 
 func (c *ntmConfig) getConfigFile() string {
 	if c.configFile == "" {
-		return "datadog.yaml"
+		for _, path := range c.configPaths {
+			configFilePath := filepath.Join(path, c.configName+".yaml")
+			if _, err := os.Stat(configFilePath); err == nil {
+				return configFilePath
```
Comment (dustmop) [inline pkg/config/nodetreemodel/read_config_file.go:26]: How about assigning to `c.configFile` here and breaking out of the loop. That way the caller doesn't need to do it.

File: pkg/config/nodetreemodel/read_config_file.go [hunk 7]
```diff
@@ -19,7 +20,12 @@ import (
 
 func (c *ntmConfig) getConfigFile() string {
 	if c.configFile == "" {
-		return "datadog.yaml"
+		for _, path := range c.configPaths {
+			configFilePath := filepath.Join(path, c.configName+".yaml")
```
Comment (dustmop) [inline pkg/config/nodetreemodel/read_config_file.go:24]: What if `c.configName` is not set? Should it default to `"datadog"` in that case?

File: pkg/config/nodetreemodel/config.go [hunk 8]
```diff
@@ -420,16 +464,7 @@ func (c *ntmConfig) AllKeysLowercased() []string {
 	return slices.Clone(c.allSettings)
 }
 
-func (c *ntmConfig) leafAtPath(key string) LeafNode {
-	return c.leafAtPathFromNode(key, c.root)
-}
-
 func (c *ntmConfig) leafAtPathFromNode(key string, curr Node) LeafNode {
```
Comment (dustmop) [inline pkg/config/nodetreemodel/config.go:467]: Is this called anywhere using a value for `curr` that isn't equal to `c.root`?

File: pkg/config/nodetreemodel/config_test.go [hunk 9]
```diff
@@ -326,12 +326,14 @@ func TestIsSet(t *testing.T) {
 	cfg := NewConfig("test", "TEST", nil)
 	cfg.SetDefault("a", 0)
 	cfg.SetDefault("b", 0)
+	cfg.SetKnown("c")
```
Comment (dustmop) [inline pkg/config/nodetreemodel/config_test.go:329]: How about adding a check for `cfg.IsKnown` for completeness.

File: pkg/config/nodetreemodel/config.go [hunk 10]
```diff
@@ -131,11 +143,13 @@ func (c *ntmConfig) OnUpdate(callback model.NotificationReceiver) {
 	c.notificationReceivers = append(c.notificationReceivers, callback)
 }
 
-func (c *ntmConfig) setDefault(key string, value interface{}) {
+func (c *ntmConfig) addToSchema(key string, source model.Source) {
 	parts := splitKey(key)
 	// TODO: Ensure that for default tree, setting nil to a node will not override
```
Comment (dustmop) [inline pkg/config/nodetreemodel/config.go:148]: I think this comment is unnecessary, it's about the default tree but this method is adding to the schema.

File: pkg/config/nodetreemodel/config.go [hunk 11]
```diff
@@ -51,6 +51,18 @@ type ntmConfig struct {
 	// Bellow are all the different configuration layers. Each layers represents a source for our configuration.
 	// They are merge into the 'root' tree following order of importance (see pkg/model/viper.go:sourcesPriority).
 
+	// schema holds all the settings with or without value. Settings are added to the schema through BindEnv and
```
Comment (dustmop) [inline pkg/config/nodetreemodel/config.go:54]: Is this table an accurate summation?

```
           +---------+------------+
           | BindEnv | SetDefault |
           +---------+------------+
AllKeys    | yes     | yes        |
AlSettings | -       | yes        |
Get        | *       | yes        |
schema     | yes     | yes        |
defaults   | -       | yes        |

*: Yes, but if and only if an env var is set with a value
```

File: pkg/config/nodetreemodel/read_config_file.go (hunk 7 above)
Comment (hush-hush) [reply pkg/config/nodetreemodel/read_config_file.go:24]: Good catch, I think initializing `configName` in `NewConfig` is better (I updated the PR).

File: pkg/config/nodetreemodel/config.go (hunk 8 above)
Comment (hush-hush) [reply pkg/config/nodetreemodel/config.go:467]: Not right now, I moved this method to the getter file where it's used.

File: pkg/config/nodetreemodel/config.go (hunk 11 above)
Comment (hush-hush) [reply pkg/config/nodetreemodel/config.go:54]: I think:

```
           +---------+------------+
           | BindEnv | SetDefault |
           +---------+------------+
AllKeys    | yes     | yes        |
AlSettings | *       | yes        |
Get        | *       | yes        |
schema     | yes     | yes        |
defaults   | -       | yes        |

*: Yes, but if and only if an env var is set with a value
```

Comment (dustmop) [reply pkg/config/nodetreemodel/config.go:54]: Right, that makes sense. Ok, I understand the difference now.

**Review by dustmop — APPROVED** [review]

**Pull request merged**